File: examples/elm-core/src/tuple.erl
-lang([erl2, st]).
-module(tuple).
-compile(export_all).

-spec pair(A, B) -> {A, B}.
pair(A, B) -> {A, B}.

-spec pair(A) -> fun((B) -> {A, B}).
pair(A) -> fun(B) -> pair(A, B) end.

-spec first({A, _}) -> A.
first({A, _}) -> A.

-spec second({_, B}) -> B.
second({_, B}) -> B.

-spec map_first(fun((A) -> X), {A, B}) -> {X, B}.
map_first(F, {A, B}) -> {F(A),B}.

-spec map_first(fun((A) -> X)) -> fun(({A, B}) -> {X, B}).
map_first(F) -> fun(P) -> map_first(F, P) end.

-spec map_second(fun((B) -> Y), {A, B}) -> {A, Y}.
map_second(F, {A, B}) -> {A,F(B)}.

-spec map_second(fun((B) -> Y)) -> fun(({A, B}) -> {A, Y}).
map_second(F) -> fun(P) -> map_second(F, P) end.

-spec map_both(fun((A) -> X), fun((B) -> Y), {A, B}) -> {X, Y}.
map_both(Fa, Fb, {A, B}) -> {Fa(A), Fb(B)}.

-spec map_both(fun((A) -> X), fun((B) -> Y)) -> fun(({A, B}) -> {X, Y}).
map_both(Fa, Fb) -> fun(P) -> map_both(Fa, Fb, P) end.

File: examples/elm-core/ir-spec/tuple.ml
let rec pair'2 : 'A * 'B -> 'A * 'B = function v_A, v_B -> (v_A, v_B)

let rec pair'1 : 'A -> 'B -> 'A * 'B = function
  | v_A -> ( function v_B -> pair'2 (v_A, v_B) )

let rec first'1 : 'A * _ -> 'A = function v_A, _ -> v_A

let rec second'1 : _ * 'B -> 'B = function _, v_B -> v_B

let rec map_first'2 : ('A -> 'X) * ('A * 'B) -> 'X * 'B = function
  | v_F, (v_A, v_B) -> (v_F v_A, v_B)

let rec map_first'1 : ('A -> 'X) -> 'A * 'B -> 'X * 'B = function
  | v_F -> ( function v_P -> map_first'2 (v_F, v_P) )

let rec map_second'2 : ('B -> 'Y) * ('A * 'B) -> 'A * 'Y = function
  | v_F, (v_A, v_B) -> (v_A, v_F v_B)

let rec map_second'1 : ('B -> 'Y) -> 'A * 'B -> 'A * 'Y = function
  | v_F -> ( function v_P -> map_second'2 (v_F, v_P) )

let rec map_both'3 : ('A -> 'X) * ('B -> 'Y) * ('A * 'B) -> 'X * 'Y = function
  | v_Fa, v_Fb, (v_A, v_B) -> (v_Fa v_A, v_Fb v_B)

let rec map_both'2 : ('A -> 'X) * ('B -> 'Y) -> 'A * 'B -> 'X * 'Y = function
  | v_Fa, v_Fb -> ( function v_P -> map_both'3 (v_Fa, v_Fb, v_P) )
